h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -267,34 +262,54 @@ public AutoML(Key<AutoML> key, Date startTime, AutoMLBuildSpec buildSpec) {
 
       handleDatafileParameters(buildSpec);
 
-    if (this.buildSpec.build_control.stopping_criteria.stopping_tolerance() == AUTO_STOPPING_TOLERANCE) {
-      this.buildSpec.build_control.stopping_criteria.set_default_stopping_tolerance_for_frame(this.trainingFrame);
+      handleEarlyStoppingParameters(buildSpec);
+
+      initLeaderboard(buildSpec);
+    } catch (Exception e) {
+      delete(); //cleanup potentially leaked keys
+      throw e;
+    }
+  }
+
+  private void initLeaderboard(AutoMLBuildSpec buildSpec) {
+    String sort_metric = buildSpec.input_spec.sort_metric == null ? null : buildSpec.input_spec.sort_metric.toLowerCase();
+    leaderboard = Leaderboard.getOrMake(projectName(), eventLog, this.leaderboardFrame, sort_metric);
+  }
+
+  private void handleReproducibilityParameters(AutoMLBuildSpec buildSpec) {
+    eventLog().info(Stage.Workflow, "Build control seed: " + buildSpec.build_control.stopping_criteria.seed() +
+            (buildSpec.build_control.stopping_criteria.seed() == -1 ? " (random)" : ""));
+  }
+
+  private void handleEarlyStoppingParameters(AutoMLBuildSpec buildSpec) {
+    if (buildSpec.build_control.stopping_criteria.stopping_tolerance() == AUTO_STOPPING_TOLERANCE) {
+      buildSpec.build_control.stopping_criteria.set_default_stopping_tolerance_for_frame(this.trainingFrame);
       eventLog().info(Stage.Workflow, "Setting stopping tolerance adaptively based on the training frame: " +
-              this.buildSpec.build_control.stopping_criteria.stopping_tolerance());
+              buildSpec.build_control.stopping_criteria.stopping_tolerance());
     } else {
-      eventLog().info(Stage.Workflow, "Stopping tolerance set by the user: " + this.buildSpec.build_control.stopping_criteria.stopping_tolerance());
+      eventLog().info(Stage.Workflow, "Stopping tolerance set by the user: " + buildSpec.build_control.stopping_criteria.stopping_tolerance());
       double default_tolerance = AutoMLBuildSpec.AutoMLStoppingCriteria.default_stopping_tolerance_for_frame(this.trainingFrame);
-      if (this.buildSpec.build_control.stopping_criteria.stopping_tolerance() < 0.7 * default_tolerance){
+      if (buildSpec.build_control.stopping_criteria.stopping_tolerance() < 0.7 * default_tolerance){
         eventLog().warn(Stage.Workflow, "Stopping tolerance set by the user is < 70% of the recommended default of " + default_tolerance + ", so models may take a long time to converge or may not converge at all.");
       }
     }
+  }
 
-      String sort_metric = buildSpec.input_spec.sort_metric == null ? null : buildSpec.input_spec.sort_metric.toLowerCase();
-      leaderboard = Leaderboard.getOrMake(projectName(), eventLog, this.leaderboardFrame, sort_metric);
-    } catch (Exception e) {
-      delete(); //cleanup potentially leaked keys
-      throw e;
+  private void handleCVParameters(AutoMLBuildSpec buildSpec) {
+    if (null != buildSpec.input_spec.fold_column && 5 != buildSpec.build_control.nfolds) {
```

Comment:
@seb-h2o Please help me to understand whether we need this block. During merge with current state of the AutoML I have not found validation for this scenario. Should we remove it?

Reply:
@deil87 you just added it, right? 
I don't see why it would be necessary: either user provides a `fold_column` and we use it (it takes precedence over `nfolds`, and we print the warning message below saying that nfolds value is ignored) or we use `nfolds`, possibly the default value of 5.
Besides, checking against a non-trivial hardcoded value is always suspicious.

I'd remove it

Reply:
@ledell if we really want to prevent the user from doing this, this is the kind of validation that should be done on client side (Py, R, schemas): setting both is harmless as soon as we know which one takes precedence, and imo, business logic should only validate against parameters that could fail the autoML process.

Reply:
besides, even if I would not raise an error for this in H2O models, at least the error is consistent: it will always raise an error if `nfolds` > 0 and `fold_column` is set.
But for AutoML, by doing this validation in business logic, it would raise an error if setting `nfolds=6, fold_column='folds'` but not if setting `nfolds=5, fold_column='folds'` which is totally weird: only clients can handle this correctly.

Reply:
@ledell I think that AutoML plays role of a client which is calling base models...that is why we do some extra validation there. If we know intentions... then we can of course pass parameters down in a right way (with warning). Though warnings are difficult to notice. But in case `nfolds=6` and `fold_column='folds'` we should fail I would say as there is no way to infer intention. Anyway, I think discussing could be moved somewhere else as topic is not directly related to this PR

Reply:
@ledell well, not on my page I guess: if I had to change something, I would change the way `ModelBuilder` validates those params, not `AutoML`.
There's an implicit rule in software: you don't fail if it doesn't prevent it from working normally or if you can easily and non-ambiguously recover from a situation, we even invented "warnings" for the later.
The question is not about inferring intentions, the code de facto gives precedence to `fold_column` (even in H2O models), we should just inform user with warnings and makes it clear in documentation. Failing for ignored params is a step towards painful API (when you want to change a param, you have to change 2 instead of 1), whereas precedence+warnings is lenient and makes is more flexible.